File: day1/main.py
import time

import numpy as np


def part_one():
    with open("input.txt", "r") as fp:
        lines = fp.readlines()

    pos = [50]
    for line in lines:
        sign = 1 if line[0] == "R" else -1
        shift = int(line[1:-1])
        pos.append(sign * shift)

    cumsum_pos = np.cumsum(pos)

    print("Stops at zero:", sum(cumsum_pos % 100 == 0))


def part_two():
    with open("input.txt", "r") as fp:
        lines = fp.readlines()

    # shifts = [50]
    pos = [50]
    at_zero_pos = 0
    within_full_shifts = 0
    through_zero = 0
    for line in lines:
        sign = 1 if line[0] == "R" else -1
        shift = int(line[1:-1])
        dividend = shift // 100
        remainder = shift % 100

        within_full_shifts += dividend

        possible_new_position = pos[-1] + (sign * remainder)
        new_position = possible_new_position % 100

        if pos[-1] != 0:
            if new_position == 0:
                at_zero_pos += 1
            elif not new_position == possible_new_position:
                through_zero += 1

        # shifts.append(sign * shift)
        pos.append(new_position)

    overall_zeros = at_zero_pos + within_full_shifts + through_zero

    # shifts = np.array(shifts)

    # div, rem = np.divmod(shifts, 100)
    #
    # no_full_shifts_pos = rem - 100 * (div < 0)

    # full_shifts = np.sum(np.abs(div)) - np.sum(div < 0)
    #
    # passed_through_zero = sum(
    #     np.not_equal(
    #         np.diff(np.cumsum(no_full_shifts_pos)),
    #         np.diff(np.remainder(np.cumsum(no_full_shifts_pos), 100)),
    #     )
    # )
    # print("Passed through zero within a step:", full_shifts)
    # print("Passed through zero from one step to another:", passed_through_zero)
    #
    # print("Overall:", full_shifts + passed_through_zero)

    print(f"Overall zeros: {overall_zeros}")


def main():
    print("Day 1 solutions:\n# First part")
    start_time = time.time()
    part_one()
    print(f"Part one took {(time.time() - start_time) * 1e3:.6f} miliseconds\n")

    print("# Second part")
    start_time = time.time()
    part_two()
    print(f"Part two took {(time.time() - start_time) * 1e3:.6f} miliseconds")


if __name__ == "__main__":
    main()

File: day1/src/main.rs
use std::fs;

fn part_one() {
    let lines = fs::read_to_string("input.txt").unwrap();
    let mut cumsum_position: Vec<i32> = Vec::with_capacity(lines.lines().count() + 1);
    cumsum_position.push(50);

    for (i, line) in lines.lines().enumerate() {
        let sign: i32 = if line.starts_with('L') { -1 } else { 1 };
        let shift: i32 = line[1..].parse().unwrap();
        cumsum_position.push(cumsum_position[i] + (sign * shift));
    }

    let rem_of_zero_pos = cumsum_position
        .iter()
        .filter(|x| x.rem_euclid(100) == 0)
        .count();

    println!("Stops at zero: {}", rem_of_zero_pos);
}

fn part_two() {
    let lines = fs::read_to_string("input.txt").unwrap();
    let mut full_shifts = 0;
    let mut last_position = 50;
    let mut at_zero_pos = 0;
    let mut through_zero = 0;
    for line in lines.lines() {
        let sign: i32 = if line.starts_with('L') { -1 } else { 1 };
        let shift: i32 = line[1..].parse().unwrap();
        let remainder = shift.rem_euclid(100);

        full_shifts += shift.div_euclid(100);

        let possible_new_position = last_position + (sign * remainder);
        let new_position = possible_new_position.rem_euclid(100);

        if last_position != 0 {
            if new_position == 0 {
                at_zero_pos += 1
            } else if new_position != possible_new_position {
                through_zero += 1
            }
        }
        last_position = new_position;
    }

    // println!("Full shifts: {}", full_shifts);
    // println!("At zero after step: {}", at_zero_pos);
    // println!("At zero during step: {}", through_zero);
    println!(
        "Overall zeros: {}",
        full_shifts + at_zero_pos + through_zero
    );
}

fn main() {
    println!("Day 1 solutions:");
    println!("# First part");
    let start_time = std::time::Instant::now();
    part_one();
    let duration = start_time.elapsed();
    println!("First part took: {:?}", duration);

    println!("\n# Second part");
    let start_time = std::time::Instant::now();
    part_two();
    let duration = start_time.elapsed();
    println!("Second part took: {:?}", duration);
}
